### DISCUSSION TITLE
init Ua Client With SIMATIC NET PC Software V15 and V13

### DISCUSSION THREAD

**Discussion by liufj** [eclipse-milo/milo#741] [General]
with version 0.2.1 ,
I can init uaClient(endpoint,ApplicationName,ApplicationUri) with SIMATIC NET PC Software V13  ,but can not init ua Client  with SIMATIC NET PC Software V15,
How can I initClient with  SIMATIC NET PC Software V15 ?  thanks!

**Comment by kevinherron** [comment]
Sorry but you’ll need to provide much more information about why it’s not working for you and what you’re trying to do. Any stack traces or error messages etc...

**Reply by liufj** [reply]
this code is working with SIMATIC NET PC Software V13, when I use SIMATIC NET PC Software V15 the  client return null.
	
     private OpcUaClient initClient(String endpointUri, String applicationUri, String applicationName)
			throws Exception {
		logger.info("startClient():" + endpointUri);//  opc.tcp://CTOS-PC:55105
		EndpointDescription[] endpoints = UaTcpStackClient.getEndpoints(endpointUri).get();
		EndpointDescription endpoint = Arrays.stream(endpoints)
				.filter(e -> e.getSecurityPolicyUri().equals(SecurityPolicy.None.getSecurityPolicyUri())).findFirst()
				.orElseThrow(() -> new Exception("no desired endpoints returned"));

		OpcUaClientConfig clientConfig = OpcUaClientConfig.builder()
				.setApplicationName(LocalizedText.english(applicationName))// "SimaticNet OpcUaServer"
				.setApplicationUri(applicationUri)// "urn:Siemens:SimaticNet:OpcUaServer:S7OPT"
				.setEndpoint(endpoint) // .setRequestTimeout(uint(60000))
				.build();
		OpcUaClient _client = new OpcUaClient(clientConfig);
		return _client;
	}

**Reply by kevinherron** [reply]
What is returning null? You’re not really making much sense. You don’t need to set those application name and URI values to match the server either; those are meant to identify your client.

Is the `endpoints` list empty or null? Can your machine resolve the address “CTOS-PC”?

**Reply by liufj** [reply]
it can  resolve the address “CTOS-PC” ,and can read node and write node,  when I use SIMATIC NET PC Software V15( OPC Server)  ,new OpcUaClient(clientConfig) is null.

**Reply by kevinherron** [reply]
I think you are misunderstanding what is happening. `new OpcUaClient(clientConfig)` can’t be null. It either gets created or possibly some Exception is thrown. You can’t successfully call a constructor for some object and end up with a `null` value.

**Reply by liufj** [reply]
Thank you for helping me lock in the wrong points,I'll go for further tests next week.
